File: frontend/src/lib/validation.js
import { BROKER_CODES } from './brokers.js';

const SYMBOL_RE = /^[A-Z0-9]{2,10}$/;
const ACCOUNT_RE = /^[A-Z0-9_.-]{1,32}$/;
const MAX_NOTE = 500;
const MIN_PRICE = 1000;
const MAX_PRICE = 10_000_000;
const MAX_MONEY = 10_000_000_000_000_000;
const MAX_SHARES = 1_000_000_000;

export class FormValidationError extends Error {
  constructor(field, en, vi, locale = 'en') {
    super(locale === 'vi' ? vi : en);
    this.field = field;
    this.code = 'CLIENT_VALIDATION';
  }
}

function number(value, field, locale, { min = null, max = null, positive = false } = {}) {
  if (String(value ?? '').trim() === '') throw new FormValidationError(field, `${field} is required.`, `${field} là bắt buộc.`, locale);
  const n = Number(value);
  if (!Number.isFinite(n)) throw new FormValidationError(field, `${field} must be a valid number.`, `${field} phải là số hợp lệ.`, locale);
  if (positive && n <= 0) throw new FormValidationError(field, `${field} must be greater than 0.`, `${field} phải lớn hơn 0.`, locale);
  if (min != null && n < min) throw new FormValidationError(field, `${field} must be at least ${min}.`, `${field} phải ít nhất ${min}.`, locale);
  if (max != null && n > max) throw new FormValidationError(field, `${field} is too large.`, `${field} vượt giới hạn cho phép.`, locale);
  return n;
}

export function parseVndMoneyInput(value, field = 'amount', locale = 'en') {
  const original = String(value ?? '').trim();
  if (!original) throw new FormValidationError(field, `${field} is required.`, `${field} là bắt buộc.`, locale);
  let text = original.toLowerCase().replace(/\s+/g, '');
  let multiplier = 1;
  let hasSuffix = false;
  if (/(triệu|tr|m)$/.test(text)) { text = text.replace(/(triệu|tr|m)$/, ''); multiplier = 1_000_000; hasSuffix = true; }
  if (!text) throw new FormValidationError(field, 'Enter a valid VND amount.', 'Nhập số tiền VND hợp lệ.', locale);
  let normalized;
  if (hasSuffix) {
    if (!/^\d+(?:[.,]\d+)?$/.test(text)) throw new FormValidationError(field, 'Enter a valid VND amount, for example 20m or 20,000,000.', 'Nhập số tiền VND hợp lệ, ví dụ 20tr hoặc 20.000.000.', locale);
    normalized = text.replace(',', '.');
  } else if (/^\d+$/.test(text)) {
    normalized = text;
  } else {
    // Without a magnitude suffix, punctuation is a thousands separator only.
    // Require one consistent separator and exactly three digits per group so
    // malformed values such as "20..000" cannot silently become 20,000 VND.
    const grouped = text.match(/^\d{1,3}([.,])\d{3}(?:\1\d{3})*$/);
    if (!grouped) throw new FormValidationError(field, 'Enter a valid VND amount, for example 20m or 20,000,000.', 'Nhập số tiền VND hợp lệ, ví dụ 20tr hoặc 20.000.000.', locale);
    normalized = text.split(grouped[1]).join('');
  }
  const n = Number(normalized) * multiplier;
  if (!Number.isFinite(n) || n < 0 || n > MAX_MONEY) throw new FormValidationError(field, 'VND amount is outside the allowed range.', 'Số tiền VND vượt phạm vi cho phép.', locale);
  if (!Number.isInteger(n)) throw new FormValidationError(field, 'VND amount must resolve to a whole number of đồng.', 'Số tiền VND phải quy đổi thành số nguyên đồng.', locale);
  return n;
}

export function validateCashAmount(value, locale = 'en') {
  const n = parseVndMoneyInput(value, 'amount', locale);
  if (n <= 0) throw new FormValidationError('amount', 'amount must be greater than 0.', 'amount phải lớn hơn 0.', locale);
  return n;
}

export function validateCashReserveInput(value, locale = 'en') {
  if (String(value ?? '').trim() === '') throw new FormValidationError('cash_reserve', 'Enter a cash reserve. Use 0 if you intentionally want no reserve.', 'Nhập mức tiền mặt dự trữ. Dùng 0 nếu bạn chủ động không giữ dự trữ.', locale);
  return parseVndMoneyInput(value, 'cash_reserve', locale);
}

export function validateBrokerAccount(brokerValue, accountValue, locale = 'en') {
  const brokerCode = String(brokerValue || 'UNASSIGNED').trim().toUpperCase();
  if (!BROKER_CODES.has(brokerCode)) throw new FormValidationError('broker_code', 'Select a supported broker.', 'Chọn công ty chứng khoán hợp lệ.', locale);
  const accountId = String(accountValue || 'PRIMARY').trim().toUpperCase();
  if (!ACCOUNT_RE.test(accountId)) throw new FormValidationError('account_id', 'Invalid account ID.', 'Mã tài khoản không hợp lệ.', locale);
  return { broker_code: brokerCode, account_id: accountId };
}

export function validateTransactionForm(type, form, today, locale = 'en') {
  const requiredSymbol = ['POSITION_IMPORT', 'BUY', 'RIGHTS_ISSUE', 'SELL', 'STOCK_DIVIDEND', 'SPLIT', 'CASH_DIVIDEND'].includes(type);
  const requiredQuantity = ['POSITION_IMPORT', 'BUY', 'RIGHTS_ISSUE', 'SELL', 'STOCK_DIVIDEND'].includes(type);
  const requiredPrice = ['POSITION_IMPORT', 'BUY', 'RIGHTS_ISSUE', 'SELL'].includes(type);
  const requiredAmount = ['CASH_DEPOSIT', 'CASH_WITHDRAW', 'CASH_DIVIDEND', 'FEE'].includes(type);
  const isTrade = ['BUY', 'RIGHTS_ISSUE', 'SELL'].includes(type);

  if (!form.event_date) throw new FormValidationError('event_date', 'Date is required.', 'Ngày là bắt buộc.', locale);
  if (today && form.event_date > today) throw new FormValidationError('event_date', 'Future dates are not allowed.', 'Không được nhập ngày trong tương lai.', locale);

  const symbol = String(form.symbol || '').trim().toUpperCase();
  if (requiredSymbol && !SYMBOL_RE.test(symbol)) throw new FormValidationError('symbol', 'Ticker must contain 2–10 letters/digits, for example FPT.', 'Mã cổ phiếu phải gồm 2–10 ký tự chữ/số, ví dụ FPT.', locale);

  const quantity = requiredQuantity ? number(form.quantity, 'quantity', locale, { positive: true, max: MAX_SHARES }) : 0;
  const price = requiredPrice ? number(form.price, 'price', locale, { positive: true, max: MAX_PRICE }) : 0;
  if (requiredPrice && price < MIN_PRICE) throw new FormValidationError('price', 'Enter the full VND price per share, for example 72,000 instead of 72.', 'Nhập giá đầy đủ theo VND/cổ phiếu, ví dụ 72.000 thay vì 72.', locale);
  const amount = requiredAmount ? validateCashAmount(form.amount, locale) : 0;
  const ratio = type === 'SPLIT' ? number(form.ratio, 'ratio', locale, { positive: true, max: 100 }) : 0;
  const fee = isTrade && String(form.fee || '').trim() !== '' ? number(form.fee, 'fee', locale, { min: 0, max: MAX_MONEY }) : 0;
  const tax = isTrade && String(form.tax || '').trim() !== '' ? number(form.tax, 'tax', locale, { min: 0, max: MAX_MONEY }) : 0;
  if (isTrade && fee + tax > quantity * price) throw new FormValidationError('fee', 'Fee + tax cannot exceed gross trade value.', 'Phí + thuế không được lớn hơn giá trị giao dịch.', locale);

  const note = String(form.note || '').trim();
  if (note.length > MAX_NOTE) throw new FormValidationError('note', `Note must be at most ${MAX_NOTE} characters.`, `Ghi chú tối đa ${MAX_NOTE} ký tự.`, locale);

  const brokerAccount = validateBrokerAccount(form.broker_code, form.account_id, locale);
  const metadata = { ...brokerAccount };
  if (isTrade) {
    const settlement = String(form.settlement_date || '').trim();
    if (settlement && settlement < form.event_date) throw new FormValidationError('settlement_date', 'Settlement date cannot be before trade date.', 'Ngày thanh toán không được trước ngày giao dịch.', locale);
    if (settlement) metadata.settlement_date = settlement;
  }

  return {
    event_type: type,
    event_date: form.event_date,
    symbol: requiredSymbol ? symbol : undefined,
    quantity, price, amount, ratio, fee, tax, note,
    broker_code: brokerAccount.broker_code,
    account_id: brokerAccount.account_id,
    metadata,
  };
}

export function validateReferenceWeightInputs(weights, symbols, locale = 'en') {
  const clean = {};
  let total = 0;
  for (const symbol of symbols) {
    const raw = weights[symbol];
    if (String(raw ?? '').trim() === '') throw new FormValidationError(symbol, `Enter a target weight for ${symbol}.`, `Nhập tỷ trọng mục tiêu cho ${symbol}.`, locale);
    const pct = number(raw, symbol, locale, { min: 0.01, max: 100 });
    clean[symbol] = pct / 100; total += pct;
  }
  if (Math.abs(total - 100) > 0.001) throw new FormValidationError('weights', `Target weights must total 100%; current total is ${total.toFixed(2)}%.`, `Tỷ trọng mục tiêu phải cộng bằng 100%; hiện tại là ${total.toFixed(2)}%.`, locale);
  return clean;
}

File: python/portfolio/validation.py
from __future__ import annotations

import math
import re
from datetime import date

from .domain import EventType

SYMBOL_RE = re.compile(r"^[A-Z0-9]{2,10}$")
ACCOUNT_RE = re.compile(r"^[A-Z0-9_.-]{1,32}$")
BROKER_RE = re.compile(r"^[A-Z0-9_.-]{2,20}$")
KNOWN_BROKERS = {"UNASSIGNED", "TCBS", "SSI", "DNSE", "VPS", "VCBS", "HSC", "VNDIRECT", "OTHER"}
MAX_NOTE_LENGTH = 500
MAX_SHARES = 1_000_000_000.0
MAX_PRICE_VND = 10_000_000.0
MIN_EQUITY_PRICE_VND = 1_000.0
MAX_MONEY_VND = 10_000_000_000_000_000.0
MAX_RATIO = 100.0


class InputValidationError(ValueError):
    def __init__(self, code: str, message: str, field: str | None = None) -> None:
        super().__init__(message)
        self.code = code
        self.field = field
        self.message = message

    def as_dict(self) -> dict:
        return {"error": self.message, "code": self.code, "field": self.field}


def _finite_number(value, field: str, *, minimum: float | None = None, maximum: float | None = None) -> float:
    try:
        number = float(value or 0)
    except (TypeError, ValueError) as exc:
        raise InputValidationError("INVALID_NUMBER", f"{field} must be a valid number.", field) from exc
    if not math.isfinite(number):
        raise InputValidationError("INVALID_NUMBER", f"{field} must be finite.", field)
    if minimum is not None and number < minimum:
        raise InputValidationError("VALUE_TOO_SMALL", f"{field} must be at least {minimum:g}.", field)
    if maximum is not None and number > maximum:
        raise InputValidationError("VALUE_TOO_LARGE", f"{field} must not exceed {maximum:g}.", field)
    return number


def _positive(value, field: str, maximum: float) -> float:
    number = _finite_number(value, field, minimum=0, maximum=maximum)
    if number <= 0:
        raise InputValidationError("REQUIRED_POSITIVE", f"{field} must be greater than 0.", field)
    return number


def _iso_date(value, field: str) -> tuple[str, date]:
    text = str(value or "").strip()
    try:
        parsed = date.fromisoformat(text)
    except ValueError as exc:
        raise InputValidationError("INVALID_DATE", f"{field} must use YYYY-MM-DD.", field) from exc
    return text, parsed


def normalize_symbol(value) -> str | None:
    symbol = str(value or "").strip().upper()
    if not symbol:
        return None
    if not SYMBOL_RE.fullmatch(symbol):
        raise InputValidationError("INVALID_SYMBOL", "symbol must contain only 2–10 uppercase letters/digits (for example FPT).", "symbol")
    return symbol


def normalize_broker(value) -> str:
    broker = str(value or "UNASSIGNED").strip().upper()
    if not BROKER_RE.fullmatch(broker):
        raise InputValidationError("INVALID_BROKER", "broker_code may contain only letters, digits, underscore, dash and dot.", "broker_code")
    if broker not in KNOWN_BROKERS:
        raise InputValidationError("UNKNOWN_BROKER", f"Unknown broker_code {broker}. Use one of: {', '.join(sorted(KNOWN_BROKERS))}.", "broker_code")
    return broker


def normalize_account(value) -> str:
    account_id = str(value or "PRIMARY").strip().upper()
    if not ACCOUNT_RE.fullmatch(account_id):
        raise InputValidationError("INVALID_ACCOUNT_ID", "account_id may contain only letters, digits, underscore, dash and dot.", "account_id")
    return account_id


def normalize_event_payload(payload: dict, *, today: str) -> dict:
    """Validate and canonicalize one ledger event request."""
    if not isinstance(payload, dict):
        raise InputValidationError("INVALID_PAYLOAD", "Request body must be an object.")
    try:
        event_type = EventType(str(payload.get("event_type") or "").strip().upper())
    except ValueError as exc:
        raise InputValidationError("INVALID_EVENT_TYPE", "Unknown event type.", "event_type") from exc

    event_date, parsed_date = _iso_date(payload.get("event_date") or today, "event_date")
    _, today_date = _iso_date(today, "today")
    if parsed_date > today_date:
        raise InputValidationError("FUTURE_DATE", "event_date cannot be in the future.", "event_date")

    note = str(payload.get("note") or "").strip()
    if len(note) > MAX_NOTE_LENGTH:
        raise InputValidationError("NOTE_TOO_LONG", f"note must be at most {MAX_NOTE_LENGTH} characters.", "note")

    metadata_raw = payload.get("metadata") or {}
    if not isinstance(metadata_raw, dict):
        raise InputValidationError("INVALID_METADATA", "metadata must be an object.", "metadata")
    metadata = dict(metadata_raw)
    metadata["broker_code"] = normalize_broker(payload.get("broker_code") or metadata.get("broker_code"))
    metadata["account_id"] = normalize_account(payload.get("account_id") or metadata.get("account_id"))
    if event_type == EventType.SELL and metadata["broker_code"] == "UNASSIGNED":
        raise InputValidationError(
            "SELL_BROKER_REQUIRED",
            "SELL requires the custody broker. Assign the holding to DNSE, TCBS or the actual broker before selling.",
            "broker_code",
        )

    symbol_required = event_type in {EventType.POSITION_IMPORT, EventType.BUY, EventType.RIGHTS_ISSUE, EventType.SELL, EventType.CASH_DIVIDEND, EventType.STOCK_DIVIDEND, EventType.SPLIT}
    symbol = normalize_symbol(payload.get("symbol"))
    if symbol_required and not symbol:
        raise InputValidationError("SYMBOL_REQUIRED", "symbol is required for this event.", "symbol")
    if not symbol_required:
        symbol = None

    quantity = 0.0
    if event_type in {EventType.POSITION_IMPORT, EventType.BUY, EventType.RIGHTS_ISSUE, EventType.SELL, EventType.STOCK_DIVIDEND}:
        quantity = _positive(payload.get("quantity"), "quantity", MAX_SHARES)

    price = 0.0
    if event_type in {EventType.POSITION_IMPORT, EventType.BUY, EventType.RIGHTS_ISSUE, EventType.SELL}:
        price = _positive(payload.get("price"), "price", MAX_PRICE_VND)
        if price < MIN_EQUITY_PRICE_VND:
            raise InputValidationError("PRICE_UNIT_SUSPECT", "price must be full VND per share (for example 72,000, not 72).", "price")

    amount = 0.0
    if event_type in {EventType.CASH_DEPOSIT, EventType.CASH_WITHDRAW, EventType.CASH_DIVIDEND, EventType.FEE}:
        amount = _positive(payload.get("amount"), "amount", MAX_MONEY_VND)

    ratio = 0.0
    if event_type == EventType.SPLIT:
        ratio = _positive(payload.get("ratio"), "ratio", MAX_RATIO)

    fee = 0.0
    tax = 0.0
    if event_type in {EventType.BUY, EventType.SELL, EventType.CASH_DIVIDEND}:
        tax = _finite_number(payload.get("tax"), "tax", minimum=0, maximum=MAX_MONEY_VND)
    if event_type in {EventType.BUY, EventType.SELL}:
        fee = _finite_number(payload.get("fee"), "fee", minimum=0, maximum=MAX_MONEY_VND)
        gross = quantity * price
        if fee + tax > gross:
            raise InputValidationError("COSTS_EXCEED_GROSS", "fee + tax cannot exceed the gross trade value.", "fee")
        metadata["trade_date"] = event_date
        settlement_raw = metadata.get("settlement_date") or payload.get("settlement_date")
        if settlement_raw:
            settlement_text, settlement_date = _iso_date(settlement_raw, "settlement_date")
            if settlement_date < parsed_date:
                raise InputValidationError("SETTLEMENT_BEFORE_TRADE", "settlement_date cannot be before the trade date.", "settlement_date")
            metadata["settlement_date"] = settlement_text
        else:
            metadata.pop("settlement_date", None)
        metadata.pop("settlement_confirmed", None)
        metadata.pop("settlement_status", None)
    else:
        for key in ("trade_date", "settlement_date", "settlement_confirmed", "settlement_status"):
            metadata.pop(key, None)

    if event_type == EventType.CASH_DIVIDEND and tax > amount:
        raise InputValidationError("COSTS_EXCEED_GROSS", "cash-dividend tax cannot exceed the gross dividend amount.", "tax")

    return {
        "event_type": event_type,
        "event_date": event_date,
        "symbol": symbol,
        "quantity": quantity,
        "price": price,
        "fee": fee,
        "tax": tax,
        "amount": amount,
        "ratio": ratio,
        "note": note,
        "metadata": metadata,
    }


def validate_reference_weights(weights: dict, *, holdings: set[str]) -> dict[str, float]:
    if not isinstance(weights, dict):
        raise InputValidationError("INVALID_WEIGHTS", "reference weights must be an object.", "weights")
    if not weights:
        return {}
    normalized: dict[str, float] = {}
    for raw_symbol, raw_weight in weights.items():
        symbol = normalize_symbol(raw_symbol)
        if not symbol:
            raise InputValidationError("INVALID_SYMBOL", "reference weight symbol is required.", "weights")
        value = _finite_number(raw_weight, f"weight[{symbol}]", minimum=0, maximum=1)
        if value <= 0:
            raise InputValidationError("INVALID_WEIGHT", f"weight[{symbol}] must be greater than 0.", "weights")
        normalized[symbol] = value
    if set(normalized) != set(holdings):
        missing = sorted(set(holdings) - set(normalized))
        extra = sorted(set(normalized) - set(holdings))
        detail = []
        if missing:
            detail.append(f"missing: {', '.join(missing)}")
        if extra:
            detail.append(f"not held: {', '.join(extra)}")
        raise InputValidationError("WEIGHT_SYMBOL_MISMATCH", "Reference weights must cover exactly the current holdings" + (f" ({'; '.join(detail)})." if detail else "."), "weights")
    total = sum(normalized.values())
    if abs(total - 1.0) > 1e-6:
        raise InputValidationError("WEIGHTS_NOT_100", f"Reference weights must total 100%; current total is {total * 100:.2f}%.", "weights")
    return normalized


def validate_cash_reserve(value) -> float:
    if value is None or (isinstance(value, str) and not value.strip()):
        raise InputValidationError("CASH_RESERVE_REQUIRED", "cash_reserve is required; use 0 only when you explicitly want no reserve.", "cash_reserve")
    return _finite_number(value, "cash_reserve", minimum=0, maximum=MAX_MONEY_VND)